### PR TITLE
added custom tooltip snippets

### LINKED ISSUE #1471: Add custom tooltip snippets
I'm not sure if there's any use-case for users. It would probably be mostly a thing for libraries and for testing those without having to compile.
My use-case here is displaying traced calls in the tooltips as if you had eval'd all the calls inline. So if you do (run-traced (some-func 5)) I can show you the result in the some-func tooltip and do the same for anything it might have called.
Another library that would benefit from this is spec/malli where you can show the spec for a function in the tooltip of the function.

I've already started building this and have played with two approaches: 
#### hover provider
This is the most flexible and easiest to build. We can simply call all customeHoverSnippets with the file, line, column and they can provide some markdown to be shown. This way any snippet could even implement links that do something, like the clojureDocs tooltips.

#### vscode.Texteditor.decorations
This is the place the inline eval values are put. Using it has a lot of disadvantages:
- hard to implement since it's a push, not pull model and snippets are pull
- harder to clear independently
- not as extensible

This kind of depends on #1442 to be really useful, so I'll base the PR on the PR for that and add it as `customTooltipSnippets` option if this doesn't make merging harder @PEZ 

As always I'd be happy to hear any feedback.

### PR DESCRIPTION
**Pull request by Cyrik**
## What has Changed?

- customHoverSnippets option added, has the same options as customREPLCommandSnippets
- evaluateCode now returns the eval result as a string

Fixes #1471 

Usage:
settings.json: 
```
"calva.customHoverSnippets": [
    {
            "name": "always hoverin",
            "snippet": "(str $hover-text)"
    }
]
```

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [ ] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [ ] Tests
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
     - [ ] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
     - [~] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.~
- [~] Created the issue I am fixing/addressing, if it was not present.~

Ping @pez, @bpringe

### CODE REVIEW

**Comment by Cyrik** [comment]
It already works pretty well, but I'll need to first merge #1470 to have the option of not showing the eval.

I'll clean up `get-text.ts` to depend on documents not the editor in the next commit. That way the hover code can also use all the other snippet replace options. I'll also add the hover stuff into the normal snippets, so you can write snippets based on mouse position. We might think about passing along any info nrepl already found for the symbol.

Otherwise, I'm pretty happy with how this turned out!

Any comments?

**Comment by PEZ** [comment]
Looks awesome!

**Comment by Cyrik** [comment]
@PEZ I think this is done.
I'd like to clean up the whole mess of `evaluate` options, but I think we should have at least a few integration tests before we pull that apart.

**Comment by Cyrik** [comment]
I just rebased + fixed my user email, not sure why I had an old work mail in here...

**Comment by PEZ** [comment]
This is so awesome. I will now test it.

We should have docs for this too, I think. Are you going to add that, @Cyrik ?

**Comment by Cyrik** [comment]
I can add some. Do you have a specific page you want it on?
I'll also write a blog about using this + the custom snippets based on integrating it into omni-trace, but I wanted to wait until it's published. It would be hard to try out when you have to install a custom vsix version.

**Comment by PEZ** [comment]
Maybe they can go on the custom snippets page for now? Or create a new page. This is certainly a feature worth it's place in the features section of the menu.

**Comment by PEZ** [comment]
We can create a separate issue for it for now, so not blocking this PR. WDYT? (The meatballs menu of my comment has a make issue item that makes creating the issue with context easy, in case you didn't know about it.)

**Comment by Cyrik** [comment]
I'll add it to the snippets page for now and we can give it a more prominent place once it's been battle-tested for a bit?

**Comment by PEZ** [comment]
customHoverSnippets doesn't seem to be registered in package.json?

**Comment by Cyrik** [comment]
Ah, good catch. I've been using it from the config.edn file which doesn't need the registration.
I'll add that together with the docs.

**Comment by PEZ** [comment]
Adding it `.calva/config.edn` I get something to happen. Maybe I don't know what to expect. I'll try to catch you on Clojurians. =)

**Comment by Cyrik** [comment]
I'll add the $currentForm stuff for hovers tonight.

**Comment by Cyrik** [comment]
So hover- for template vars that make sense now exists and hover can also use the basic template vars. The code for that could use a little abstraction but switching camelCase to kebabCase on properties is kind of annoying, so I'll think about handling that later.

I think I'm happy to have this in dev, what do you think @PEZ ?

**Comment by PEZ** [comment]
(For anyone reading this and thinking I am not answering, @Cyrik and I are also chatting on Clojurians 😄).

**Comment by PEZ** [comment]
Tested this a bit now. Some strange behaviour still:

The first time I hover something in a new file/tab a lot of prompts are appearing in the output window, with the file's namespace. 18 prompts in my case, might be a factor of how many snippets I have defined.

The snippets appear to never be cleared from the hover, just added, making them appear many times in the hover. 

image

 I can't get snippets from JSON settings to appear.

Another thing I am struggling a bit with understanding is how to get markdown into the hover w/o quotes being added. I tried with this:

```edn
  {:name "edn hover `current-form`"
   :snippet "\"`$hover-current-form`\""}
```

If I hover a symbol `pirate-char`, the hover shows:

"`pirate-char`"

Maybe we need to treat strings special when adding to the hover?

**Comment by PEZ** [comment]
Actually the prompts seem to be as many as the times the snippets appear in the hover. It got to be more and more prompts as I hovered.

File: docs/site/custom-commands.md
Comment (bpringe) [inline docs/site/custom-commands.md]:
```suggestion
Be careful with these, since they will be executed anytime Calva displays a tooltip. So they should be fast and probably not have any side effects.
```

File: docs/site/custom-commands.md
Comment (bpringe) [inline docs/site/custom-commands.md]:
```suggestion
With this setting anything the mouse is over will also be shown inside its tooltip.
```

**Review by bpringe — CHANGES_REQUESTED** [review]

**Comment by bpringe** [comment]
I didn't thoroughly review the code; just requested a couple small docs changes. I'll leave the code review to @PEZ.

**Comment by Cyrik** [comment]
Turns out `_.remove()` returns the removed items and mutates the given array, that's not completely what I expected 😉 
Hope that was the last bug left in here.

**Comment by PEZ** [comment]
> Turns out _.remove() returns the removed items and mutates the given array

I've been using that footgun like that as well. 😄  Also `.slice()` on JS arrays behave like this.

**Review by PEZ — APPROVED** [review]
Like a charm! Thanks for your work! ❤️ 🙏 And for your patience and persistence 😄

**Comment by PEZ** [comment]
We probably should add a setting where you can allow/disallow hover snippets from certain dependencies.

**Comment by Cyrik** [comment]
> We probably should add a setting where you can allow/disallow hover snippets from certain dependencies.

Yeah, probably for custom snippets as well.

**Comment by bpringe** [comment]
I'm going to be the PR police here... 😄 

A changelog entry was not added, and the issue was not mentioned in a commit for this PR, even though that checkbox was checked.

> Referenced the issue I am fixing/addressing in a commit message for the pull request.

Just FYI, though, according to @PEZ, even though mentioning the issue in the PR description doesn't auto-close the issue, linking it to the PR manually does auto-close the issue.

PR police out...

giphy

**Comment by bpringe** [comment]
I linked the issue and added the changelog entry on dev.

**Comment by PEZ** [comment]
Haha, all this work reviewing this PR and I missed about the changelog! 😄 I'll add that in `dev`.

Regarding linking the issues in the commits. I think we should remove that checkbox and add a checkbox for the Calva team to link the issue manually before merging.

**Comment by Cyrik** [comment]
Oh yeah sorry about the changelog thing. After getting merge conflicts for it all the time I've stopped adding them at the start, but now I'm forgetting to add them at the end 😅

**Pull request merged**
